test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java

```diff
@@ -312,30 +312,115 @@ public void testUnknownField() throws IOException {
+        unknownObjectExceptionTest(validQuery);
+        for (String query : getAlternateVersions().keySet()) {
+            unknownObjectExceptionTest(query);
+        }
+    }
+
+    private void unknownObjectExceptionTest(String validQuery) throws IOException {
+        //TODO building json by concatenating strings makes the code unmaintainable, we should rewrite this test
         assertThat(validQuery, containsString("{"));
+        int level = 0;
+        boolean withinQuotes = false;
+        boolean expectedException = true;
+        int objectHoldingArbitraryContentLevel = 0;
         for (int insertionPosition = 0; insertionPosition < validQuery.length(); insertionPosition++) {
-            if (validQuery.charAt(insertionPosition) == '{') {
-                String testQuery = validQuery.substring(0, insertionPosition) + "{ \"newField\" : "
-                        + validQuery.substring(insertionPosition) + "}";
+            if (validQuery.charAt(insertionPosition) == '"') {
+                withinQuotes = withinQuotes == false;
+            } else if (withinQuotes == false && validQuery.charAt(insertionPosition) == '}') {
+                level--;
+                if (expectedException == false) {
+                    //track where we are within the object that holds arbitrary content
+                    objectHoldingArbitraryContentLevel--;
+                }
+                if (objectHoldingArbitraryContentLevel == 0) {
+                    //reset the flag once we have traversed the whole object that holds arbitrary content
+                    expectedException = true;
+                }
+            } else if (withinQuotes == false && validQuery.charAt(insertionPosition) == '{') {
+                level++;
+                if (expectedException) {
+                    int start = -1;
+                    int end = -1;
+                    for (int i = insertionPosition; i >= 0; i--) {
+                        if (validQuery.charAt(i) == '}') {
+                            break;
+                        } else if (validQuery.charAt(i) == '"') {
+                            if (end == -1) {
+                                end = i;
+                            } else if (start == -1) {
+                                start = i + 1;
+                            } else {
+                                break;
+                            }
+                        }
+                    }
+                    if (start >= 0  && end > 0) {
+                        String objectName = validQuery.substring(start, end);
+                        expectedException = getObjectsHoldingArbitraryContent().contains(objectName) == false;
+                    }
+                }
+                if (expectedException == false) {
+                    objectHoldingArbitraryContentLevel++;
+                }
+                String testQuery = validQuery.substring(0, insertionPosition) + "{ \"newField\" : ";
```

Comment:
Maybe point out that this is actually the place where we modify the valid input query by adding a new object level to it.